File: extension/src/main/java/io/holunda/camunda/bpm/data/writer/VariableScopeWriter.java
package io.holunda.camunda.bpm.data.writer;

import io.holunda.camunda.bpm.data.factory.VariableFactory;
import java.util.function.Supplier;
import org.camunda.bpm.engine.delegate.VariableScope;
import org.camunda.bpm.engine.variable.VariableMap;

/**
 * Variable scope builder allowing for fluent variable setting.
 */
public class VariableScopeWriter implements LocalVariableWriter<VariableScopeWriter> {

  private final VariableScope scope;

  /**
   * Creates a builder with provided variable map.
   * <p>The provided variables are modified by reference.</p>
   *
   * @param variables variables to work on.
   */
  public VariableScopeWriter(VariableScope variables) {
    this.scope = variables;
  }

  @Override
  public <T> VariableScopeWriter set(VariableFactory<T> factory, T value) {
    return this.set(factory, value, false);
  }

  @Override
  public <T> VariableScopeWriter set(VariableFactory<T> factory, T value, boolean isTransient) {
    factory.on(this.scope).set(value, isTransient);
    return this;
  }

  @Override
  public <T> VariableScopeWriter setLocal(VariableFactory<T> factory, T value) {
    return this.setLocal(factory, value, false);
  }

  @Override
  public <T> VariableScopeWriter setLocal(VariableFactory<T> factory, T value, boolean isTransient) {
    factory.on(this.scope).setLocal(value, isTransient);
    return this;
  }

  @Override
  public <T> VariableScopeWriter remove(VariableFactory<T> factory) {
    factory.on(this.scope).remove();
    return this;
  }


  @Override
  public <T> VariableScopeWriter removeLocal(VariableFactory<T> factory) {
    factory.on(this.scope).removeLocal();
    return this;
  }


  @Override
  public VariableMap variables() {
    return scope.getVariablesTyped();
  }

  @Override
  public VariableMap variablesLocal() {
    return scope.getVariablesLocalTyped();
  }

}

File: extension/c7/src/main/kotlin/io/holunda/camunda/bpm/data/writer/VariableScopeWriter.kt
package io.holunda.camunda.bpm.data.writer

import io.holunda.camunda.bpm.data.factory.VariableFactory
import io.holunda.camunda.bpm.data.adapter.WriteAdapter
import io.holunda.camunda.bpm.data.factory.*
import org.camunda.bpm.engine.delegate.VariableScope
import org.camunda.bpm.engine.variable.VariableMap

/**
 * Variable scope builder allowing for fluent variable setting.
 * @param scope variables to work on.
 */
class VariableScopeWriter(
  private val scope: VariableScope
) : VariableWriter<VariableScopeWriter>, C7ServiceVariableWriter<VariableScopeWriter>() {

  companion object {
    @JvmStatic
    @Suppress("UNCHECKED_CAST")
    fun <T> of(variableFactory: VariableFactory<T>, scope: VariableScope): WriteAdapter<T> = when (variableFactory) {
      is BasicVariableFactory<*> -> C7Adapters.on(variableFactory, scope)
      is ListVariableFactory<*> -> C7Adapters.on(variableFactory, scope)
      is SetVariableFactory<*> -> C7Adapters.on(variableFactory, scope)
      is MapVariableFactory<*, *> -> C7Adapters.on(variableFactory, scope)
      else -> throw IllegalArgumentException("Unsupported factory type of type ${variableFactory.javaClass}")
    } as WriteAdapter<T>
  }

  override fun <T> getWriter(variableFactory: VariableFactory<T>): WriteAdapter<T> = of(variableFactory, scope)

  override fun variables(): VariableMap {
    return scope.variablesTyped
  }

  override fun variablesLocal(): VariableMap {
    return scope.variablesLocalTyped
  }

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (other == null || javaClass != other.javaClass) return false
    val that = other as VariableScopeWriter
    return scope == that.scope
  }

  override fun hashCode(): Int {
    return scope.hashCode()
  }
}
